scripts/globals/additional_effects.lua

```diff
@@ -0,0 +1,274 @@
+        local MP = damage
+        -- Unknown what modifies this, using power directly for now
+        msgID = xi.msg.basic.ADD_EFFECT_MP_HEAL
+        attacker:addMP(MP)
+        msgParam = MP
+
+    elseif addType == procType.HP_DRAIN or (addType == procType.HPMPTP_DRAIN and math.random(1,3) == 1) then
+        damage = xi.additionalEffect.calcDamage(attacker, element, defender, damage)
+        if damage > defender:getHP() then
+            damage = defender:getHP()
+        end
+
+        msgID = xi.msg.basic.ADD_EFFECT_HP_DRAIN
+        msgParam = damage
+        defender:addHP(-damage)
+        attacker:addHP(damage)
+
+    elseif addType == procType.MP_DRAIN or (addType == procType.HPMPTP_DRAIN and math.random(1,3) == 2) then
+        damage = xi.additionalEffect.calcDamage(attacker, element, defender, damage)
+        if damage > defender:getMP() then
+            damage = defender:getMP()
+        end
+        msgID = xi.msg.basic.ADD_EFFECT_MP_DRAIN
+        msgParam = damage
+        defender:addMP(-damage)
+        attacker:addMP(damage)
+
+    elseif addType == procType.TP_DRAIN or (addType == procType.HPMPTP_DRAIN and math.random(1,3) == 3) then
+        damage = xi.additionalEffect.calcDamage(attacker, element, defender, damage)
+        if damage > defender:getTP() then
+            damage = defender:getTP()
+        end
+        msgID = xi.msg.basic.ADD_EFFECT_TP_DRAIN
+        msgParam = damage
+        defender:addTP(-damage)
+        attacker:addTP(damage)
+
+    elseif addType == procType.DISPEL then
+        local dispel = defender:dispelStatusEffect()
+        -- Resistance check should be in dispelStatusEffect() itself
+        if dispel == xi.effect.NONE then
+            return 0, 0, 0
+        else
+            msgID = xi.msg.basic.ADD_EFFECT_DISPEL
+            msgParam = dispel
+        end
+
+    elseif addType == procType.ABSORB then
+        -- Ripping off Aura Steal here
+        local resist = applyResistanceAddEffect(attacker, defender, element, 0)
+        if resist > 0.0625 then
+            local stolen = attacker:stealStatusEffect(defender)
+            msgID = xi.msg.basic.STEAL_EFFECT
+            msgParam = stolen
+        end
+
+    elseif addType == procType.SELF_BUFF then
+        if addStatus == xi.effect.TELEPORT then -- WARP
+            attacker:addStatusEffectEx(xi.effect.TELEPORT, 0, xi.teleport.id.WARP, 0, 1)
+            msgID = xi.msg.basic.ADD_EFFECT_WARP
+            msgParam = 0
+        elseif addStatus == xi.effect.BLINK then -- BLINK http://www.ffxiah.com/item/18830/gusterion
+            -- Does not stack with or replace other shadows
+            if attacker:hasStatusEffect(xi.effect.BLINK) or attacker:hasStatusEffect(xi.effect.UTSUSEMI) then
+                return 0, 0, 0
+            else
+                attacker:addStatusEffect(xi.effect.BLINK, power, 0, duration)
+                msgID = ADD_EFFECT_SELFBUFF
```

Comment:
Did you mean to reference xi.msg.basic here?

Reply:
yep thanks